config/astyaml/astyaml.go

```diff
@@ -12,79 +12,127 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package config
+package astyaml
 
 import (
 	"errors"
 	"io"
+	"strings"
 
 	yaml "github.com/ajeddeloh/yaml"
 	"github.com/coreos/ignition/config/validate"
 )
 
 var (
 	ErrNotDocumentNode = errors.New("Can only convert from document node")
+	ErrNotMappingNode  = errors.New("Tried to change the key of a node which is not a mapping node")
+	ErrKeyNotFound     = errors.New("Key to be replaced not found")
 )
 
-type yamlNode struct {
+type YamlNode struct {
+	tag string
 	key yaml.Node
 	yaml.Node
 }
 
-func fromYamlDocumentNode(n yaml.Node) (yamlNode, error) {
+func FromYamlDocumentNode(n yaml.Node) (YamlNode, error) {
 	if n.Kind != yaml.DocumentNode {
-		return yamlNode{}, ErrNotDocumentNode
+		return YamlNode{}, ErrNotDocumentNode
 	}
 
-	return yamlNode{
+	return YamlNode{
 		key:  n,
+		tag:  "yaml",
 		Node: *n.Children[0],
 	}, nil
 }
 
-func (n yamlNode) ValueLineCol(source io.ReadSeeker) (int, int, string) {
+func (n YamlNode) ValueLineCol(source io.ReadSeeker) (int, int, string) {
 	return n.Line + 1, n.Column + 1, ""
 }
 
-func (n yamlNode) KeyLineCol(source io.ReadSeeker) (int, int, string) {
+func (n YamlNode) KeyLineCol(source io.ReadSeeker) (int, int, string) {
 	return n.key.Line + 1, n.key.Column + 1, ""
 }
 
-func (n yamlNode) LiteralValue() interface{} {
+func (n YamlNode) LiteralValue() interface{} {
 	return n.Value
 }
 
-func (n yamlNode) SliceChild(index int) (validate.AstNode, bool) {
+func (n YamlNode) SliceChild(index int) (validate.AstNode, bool) {
 	if n.Kind != yaml.SequenceNode {
 		return nil, false
 	}
 	if index >= len(n.Children) {
 		return nil, false
 	}
 
-	return yamlNode{
+	return YamlNode{
 		key:  yaml.Node{},
+		tag:  n.tag,
 		Node: *n.Children[index],
 	}, true
 }
 
-func (n yamlNode) KeyValueMap() (map[string]validate.AstNode, bool) {
+func (n YamlNode) KeyValueMap() (map[string]validate.AstNode, bool) {
 	if n.Kind != yaml.MappingNode {
 		return nil, false
 	}
 
 	kvmap := map[string]validate.AstNode{}
 	for i := 0; i < len(n.Children); i += 2 {
 		key := *n.Children[i]
+		if n.tag == "json" {
+			key.Value = getIgnKeyName(key.Value)
+		}
 		value := *n.Children[i+1]
-		kvmap[key.Value] = yamlNode{
+		kvmap[key.Value] = YamlNode{
 			key:  key,
+			tag:  n.tag,
 			Node: value,
 		}
 	}
 	return kvmap, true
 }
 
-func (n yamlNode) Tag() string {
-	return "yaml"
+// ChangeKey replaces the oldkey with a new key/value pair. Useful for patching
+// up a tree parsed from yaml but then used for validating an ignition structure
+func (n *YamlNode) ChangeKey(oldKeyName, newKeyName string, newValue YamlNode) error {
+	if n.Kind != yaml.MappingNode {
+		return ErrNotMappingNode
+	}
+	for i := 0; i < len(n.Children); i += 2 {
```

Comment:
go-yaml parses a mapping type into an array of interleaved key and value nodes, so even nodes are the keys and odd nodes the values.